### PR TITLE
Add new community pipeline for 'Adaptive Mask Inpainting', introduced in [ECCV2024] ComA

### PR DESCRIPTION
**Pull request by jellyheadandrew**
Add new community pipeline for 'Adaptive Mask Inpainting' introduced in **[ECCV2024] Beyond the Contact: Discovering Comprehensive Affordance for 3D Objects from Pre-trained 2D Diffusion Models**.

# What does this PR do?

This PR implements the 'Adaptive Mask Inpainting' algorithm introduced in **[ECCV2024] Beyond the Contact: Discovering Comprehensive Affordance for 3D Objects from Pre-trained 2D Diffusion Models**.

The code is borrowed from the author's repository (https://github.com/snuvclab/coma), changed slightly to represent the demo usage.

In a nutshell, this pipeline provides a way to insert human inside the scene image without altering the background, by inpainting with adapting mask.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by asomoza** [comment]
Hi @jellyheadandrew, this seems really interesting, can you please post some example generations of this?

**Comment by jellyheadandrew** [comment]
Yeah sure! 

https://github.com/user-attachments/assets/7950963
image (2)
a-ac92-47c1-9d01-bd17bda7e62d

File: adaptive_mask_inpainting_example.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    """
+    
+    adaptive_mask_model = PointRendPredictor(
+        pointrend_thres=0.2, 
+        device="cuda" if torch.cuda.is_available() else "cpu", 
+        use_visualizer=use_visualizer,
+        config_pth="pointrend_rcnn_R_50_FPN_3x_coco.yaml",
+        weights_pth="model_final_edd263.pkl",
+    )
+    pipeline.register_adaptive_mask_model(adaptive_mask_model)
```
Comment (asomoza) [inline adaptive_mask_inpainting_example.py:132]: you can probably remove this from here and add it as code inside the pipeline to make it easier for the user. But it's your decision, if you do you should probably download the files inside the pipeline too then.

File: adaptive_mask_inpainting_example.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            ),
+            dilate_kernel=np.ones((3, 3), dtype=np.uint8),
+            provoke_scheduler=ProvokeScheduler(
+                num_inference_steps=num_steps,
+                schedule=list(range(2, 10 + 1, 2)) + list(range(12, 40 + 1, 2)) + [45],
+                is_zero_indexing=False,
+            ),
+        )
+    )
+    pipeline.register_adaptive_mask_settings(adaptive_mask_settings)
```
Comment (asomoza) [inline adaptive_mask_inpainting_example.py:152]: same as before

**Review by asomoza — COMMENTED** [review]
Thanks, if you're still interested in adding it to diffusers I left a couple of comments.

File: adaptive_mask_inpainting_example.py (hunk 2 above)
Comment (jellyheadandrew) [reply adaptive_mask_inpainting_example.py:152]: Thanks for the thorough feedback! I've applied them, and will PR again.

(Please check minor questions above)

File: adaptive_mask_inpainting_example.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+    num_steps = 50
+    
+    # Scheduler
+    scheduler = DDIMScheduler(
+        beta_start=0.00085, 
+        beta_end=0.012, 
+        beta_schedule="scaled_linear", 
+        clip_sample=False, 
+        set_alpha_to_one=False
```
[the remark this answers is not in the document]
Comment (jellyheadandrew) [reply adaptive_mask_inpainting_example.py:98]: Sorry, I'm kinda new to contributing to huggingface. What do you mean by changing the scheduler? Do you mean changing the config?

[the remark this answers is not in the document]
Comment (asomoza) [reply adaptive_mask_inpainting_example.py:98]: oh, no problem, normally we can just change the scheduler with any other scheduler just using it like this:

```python
from diffusers import DDIMScheduler
....
pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
````

which is cleaner and simpler.

[the remark this answers is not in the document]
Comment (asomoza) [reply adaptive_mask_inpainting_example.py:98]: and if it has a different config (which doesn't seem the case here) you can just add the changed param as an additional kwarg

**Comment by jellyheadandrew** [comment]
Hi, I have updated the files based on your feedbacks. Waiting for more feedbacks, or a merge!

**Comment by jellyheadandrew** [comment]
Hi @asomoza , I have updated the files based on your feedbacks. Waiting for more feedbacks, or a merge!

**Comment by asomoza** [comment]
Hi @jellyheadandrew sorry it took me so long to get back to you. Let's finish this so we can merge it. 

First can you please merge it to the main branch, I can see some couple of outdated imports that will probably fail. 

Also can you run `make style` and `make quality`.

**Comment by jellyheadandrew** [comment]
@asomoza Hi, I've updated my branch! Could you check

**Comment by asomoza** [comment]
@jellyheadandrew thanks, you merged main but we still need the `make style` and `make quality` like it's described in the [contributing guide](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md#how-to-open-a-pr). Without that the tests will fail, also you have some unused imports that will be fixed.

**Comment by jellyheadandrew** [comment]
Thanks @asomoza , I've finished `make style` and `make quality`.

**Comment by yiyixuxu** [comment]
@jellyheadandrew merged! thanks for your contribution!

**Pull request merged**